pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -2271,48 +2278,123 @@ private void initDeadLetterProducerIfNeeded() {
             createProducerLock.writeLock().lock();
             try {
                 if (deadLetterProducer == null) {
-                    deadLetterProducer =
-                            ((ProducerBuilderImpl<byte[]>) client.newProducer(Schema.AUTO_PRODUCE_BYTES(schema)))
-                                    .initialSubscriptionName(this.deadLetterPolicy.getInitialSubscriptionName())
-                                    .topic(this.deadLetterPolicy.getDeadLetterTopic())
-                                    .producerName(String.format("%s-%s-%s-%s-DLQ", this.topicName, this.subscription,
-                                            this.consumerName, RandomStringUtils.randomAlphanumeric(5)))
-                                    .blockIfQueueFull(false)
-                                    .enableBatching(false)
-                                    .enableChunking(true)
-                                    .createAsync();
-                    deadLetterProducer.thenAccept(dlqProducer -> {
-                        stats.setDeadLetterProducerStats(dlqProducer.getStats());
-                    });
+                    deadLetterProducer = createProducerWithBackOff(() -> {
+                        CompletableFuture<Producer<byte[]>> newProducer =
+                                ((ProducerBuilderImpl<byte[]>) client.newProducer(Schema.AUTO_PRODUCE_BYTES(schema)))
+                                        .initialSubscriptionName(this.deadLetterPolicy.getInitialSubscriptionName())
+                                        .topic(this.deadLetterPolicy.getDeadLetterTopic())
+                                        .producerName(
+                                                String.format("%s-%s-%s-%s-DLQ", this.topicName, this.subscription,
+                                                        this.consumerName, RandomStringUtils.randomAlphanumeric(5)))
+                                        .blockIfQueueFull(false)
+                                        .enableBatching(false)
+                                        .enableChunking(true)
+                                        .createAsync();
+                        newProducer.thenAccept(dlqProducer -> {
+                            stats.setDeadLetterProducerStats(dlqProducer.getStats());
+                        });
```

Comment:
Is it possible to move all logic about `closeProducer` and `set retryLetterProducerFailureCount = 0` to here?

Includes the retryProducer as well.

```
                        newProducer.thenAccept(dlqProducer -> {
                            stats.setDeadLetterProducerStats(dlqProducer.getStats());
                            deadLetterProducerFailureCount = 0;
                        }).exceptionally(ex -> {
                            log.error("Dead letter producer exception with topic: {}",
                                    deadLetterPolicy.getDeadLetterTopic(), ex);
                            closeDeadLetterProducerAfterException();
                            return null;
                        });
```

Reply:
I tried that too. In that case there would be no backoff. For example in the test case, the producer gets successfully, but the message sending fails. That's why failure count is reseted only after successful message sending. 
However, since I'm thinking of revisiting the logic based on the previous comment, this logic would be obsolete in that case.